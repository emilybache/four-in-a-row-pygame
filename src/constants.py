# CONSTANTS, Game Logic #
RED, YELLOW, EMPTY = range(3)
COLUMNS = 7
ROWS = 6
DROP_DELAY_MS = 250

# CONSTANTS, Graphics #
WIDTH, HEIGHT = 1100, 800
CENTER = (WIDTH // 2, HEIGHT // 2)
CENTER_X, CENTER_Y = CENTER
SCREENDIM = (WIDTH, HEIGHT)
DISC_DIAMETER = 90
DISC_RADIUS = DISC_DIAMETER // 2
BIG_TEXT = 45
MID_TEXT = 30
BOARD_WIDTH = DISC_DIAMETER * COLUMNS
BOARD_HEIGHT = DISC_DIAMETER * ROWS
BOARD_LEFT = (WIDTH - BOARD_WIDTH) // 2
BOARD_RIGHT = (WIDTH + BOARD_WIDTH) // 2
FPS = 60


class Color:
    WHITE = (255, 255, 255)
    BLACK = (0, 0, 0)
    RED = (255, 0, 0)
    GREEN = (0, 255, 0)
    BLUE = (0, 0, 255)
    YELLOW = (255, 255, 0)