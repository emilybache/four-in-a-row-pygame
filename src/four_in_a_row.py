from collections import defaultdict
import pygame

# CONSTANTS #
DISC_SIZE = 90
RED, YELLOW, EMPTY = range(3)
WIDTH, HEIGHT = 1100, 600
CENTER = (WIDTH // 2, HEIGHT // 2)
COLUMNS = 7
ROWS = 6


class Color:
    WHITE = (255, 255, 255)
    BLACK = (0, 0, 0)
    RED = (255, 0, 0)
    GREEN = (0, 255, 0)
    BLUE = (0, 0, 255)
    YELLOW = (255, 255, 0)


# STATES #

class GameState(object):
    def __init__(self):
        self.board = defaultdict(lambda: EMPTY)
        self.whos_turn_is_it = RED

    def whos_turn(self):
        return print_color(self.whos_turn_is_it)


class GameOverState(object):
    def __init__(self, winner):
        self.winner = winner


class StartScreenState(object):
    pass


# MESSAGES #

class LeftMouseClickAt(object):
    def __init__(self, pos):
        self.pos = pos


class ColumnWasClicked(object):
    def __init__(self, column):
        self.column = column


# FUNCTIONS #

def update(model, msg):
    if isinstance(model, GameState):
        if isinstance(msg, ColumnWasClicked):
            column = msg.column
            model.board = place_brick(model.board, model.whos_turn_is_it, column)
            model.whos_turn_is_it = (model.whos_turn_is_it + 1) % 2
            won = check_winning_state(model.board, RED)
            if won:
                return GameOverState(winner=RED)
            won = check_winning_state(model.board, YELLOW)
            if won:
                return GameOverState(winner=YELLOW)
    if isinstance(model, StartScreenState):
        if isinstance(msg, LeftMouseClickAt):
            return GameState()

    return model


def place_brick(board, color, column):
    log(f"Placing brick color {print_color(color)} in column {column}")
    for i in range(ROWS):
        y = ROWS - i - 1
        if board[(column, y)] == EMPTY:
            board[(column, y)] = color
            log("Board state is now:")
            log(board_to_string(board))
            break
    return board


def board_to_string(board):
    board_string = ''
    symbols = {
        EMPTY: 'O',
        RED: 'R',
        YELLOW: 'Y'
    }
    board_string += "0 1 2 3 4 5 6\n"
    board_string += "-------------\n"
    for y in range(6):
        for x in range(7):
            pos = (x, y)
            cell = board[pos]
            board_string += symbols[cell] + ' '
        board_string += '\n'
    return board_string


def print_color(color):
    return 'red' if color == RED else 'yellow'


def check_winning_state(board, color):
    for (x, y) in all_positions():
        for dir in [(0, 1), (1, 0), (1, 1), (-1, 1)]:
            cells = extract(board, (x, y), dir)
            if all(cell == color for cell in cells):
                log(f"Found 4-in-a-row at {x, y} dir {dir}")
                return True
    return False


def extract(board, pos, dir):
    cells = []
    for i in range(4):
        p = (pos[0] + dir[0] * i, pos[1] + dir[1] * i)
        cells.append(board[p])
    return cells


def all_positions():
    for y in range(ROWS):
        for x in range(COLUMNS):
            yield (x, y)


# PyGame drawing wrapper
class DrawingAPI:
    def __init__(self, screen):
        self.screen = screen
        self.font_name = pygame.font.match_font('arial')

    def draw_rectangle(self, center, size, color):
        log(f"Drawing rectangle center {center} size {size} color {color}")
        r = pygame.Rect(0, 0, *size)
        r.center = center
        pygame.draw.rect(self.screen, color, r)

    def draw_disc(self, center, size, color):
        log(f'Drawing a disc center {center} size {size} color {color}')
        pygame.draw.circle(self.screen, color, center, size, size)

    def draw_text(self, center, text, size, color):
        log(f"Drawing text '{text}' at {center} color {color} size {size}")
        font = pygame.font.Font(self.font_name, size)
        text_surface = font.render(text, True, color)
        text_rect = text_surface.get_rect()
        text_rect.center = center
        self.screen.blit(text_surface, text_rect)


def view(model, api):
    if isinstance(model, GameState):
        api.draw_rectangle(CENTER, (WIDTH, HEIGHT), Color.BLACK)
        api.draw_rectangle(CENTER, (DISC_SIZE * COLUMNS + 20, DISC_SIZE * ROWS + 20), Color.BLUE)
        for (x, y) in all_positions():
            value = model.board[(x, y)]
            color = Color.BLACK
            if value != EMPTY:
                color = rgb_from_color(value)
            x0 = int(WIDTH / 2 - COLUMNS * DISC_SIZE / 2 + x * DISC_SIZE + DISC_SIZE // 2)
            y0 = int(HEIGHT / 2 - ROWS * DISC_SIZE / 2 + y * DISC_SIZE + DISC_SIZE // 2)
            pos = (x0, y0)
            api.draw_disc(pos, DISC_SIZE // 2, color)
        api.draw_text((WIDTH // 2, 20), f"{model.whos_turn().title()} to place disc", 30, Color.WHITE)
    if isinstance(model, GameOverState):
        api.draw_rectangle(CENTER, (WIDTH, HEIGHT), Color.BLUE)
        api.draw_text((CENTER[0], CENTER[1] - 45), f"GAME OVER", 45, Color.WHITE)
        api.draw_text((CENTER[0], CENTER[1] + 45),
                      f"{print_color(model.winner)} won!".upper(),
                      45,
                      rgb_from_color(model.winner)
                      )
    if isinstance(model, StartScreenState):
        api.draw_rectangle(CENTER, (WIDTH, HEIGHT), Color.GREEN)
        api.draw_text((CENTER[0], CENTER[1] - 45), "FOUR-IN-A-ROW", 45, Color.WHITE)
        api.draw_text((CENTER[0], CENTER[1] + 45), "Click left mouse button to play!", 45, Color.WHITE)


def rgb_from_color(color):
    return Color.RED if color == RED else Color.YELLOW


# MAIN PROGRAM #


def log(msg):
    print(msg)


def print_model(model):
    state_string = model.__class__.__name__ + '\n'
    if isinstance(model, StartScreenState):
        pass
    if isinstance(model, GameOverState):
        state_string += f'{print_color(model.winner).title()} won.\n'
    if isinstance(model, GameState):
        state_string += f'It is {model.whos_turn()}s turn.\n'
        state_string += board_to_string(model.board)
    return state_string


def mainloop(drawing_api):
    model = StartScreenState()
    i = 0
    view(model, drawing_api)
    while True:
        # Translate low level events to domain events
        msgs = []
        ev = pygame.event.poll()
        if ev.type == pygame.MOUSEBUTTONDOWN:
            print(f"Got click {ev}")
            pos = ev.pos
            button = ev.button
            if button == 1:
                msgs.append(LeftMouseClickAt(pos))
        if ev.type == pygame.KEYDOWN:
            if ev.key == pygame.K_RETURN:
                msgs.append(ColumnWasClicked(i % 2))
                i += 1
            elif ev.key == pygame.K_q:
                break
        elif ev.type == pygame.QUIT:
            break

        # Handle events to update state
        old_model_repr = print_model(model)
        model = run_messages(model, msgs)

        # Display current model
        if old_model_repr != print_model(model):
            view(model, api)

        pygame.display.update()


def run_messages(model, msgs):
    for msg in msgs:
        model = update(model, msg)
    return model


if __name__ == '__main__':
    pygame.init()
    screen = pygame.display.set_mode((WIDTH, HEIGHT))
    pygame.display.set_caption("Four in a row")
    api = DrawingAPI(screen)
    mainloop(api)
    pygame.quit()
