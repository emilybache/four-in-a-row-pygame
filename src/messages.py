from collections import namedtuple

LeftMouseDownAt = namedtuple('LeftMouseDownAt', 'pos')
LeftMouseUpAt = namedtuple('LeftMouseUpAt', 'pos')
ColumnWasClicked = namedtuple('ColumnWasClicked', 'column')
MouseMovedTo = namedtuple('MouseMovedTo', 'pos')
Tick = namedtuple('Tick', 'time')