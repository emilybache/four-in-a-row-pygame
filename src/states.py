from collections import defaultdict, namedtuple

from src.constants import CENTER, RED, EMPTY


class StartScreenState(object):
    def __init__(self):
        self.time = 0
        self.music_playing = False


class GameState(object):
    def __init__(self):
        self.board = empty_board()
        self.whos_turn_is_it = RED
        self.mouse_pos = CENTER
        self.mouse_down_time = None
        self.time = None


def empty_board():
    return defaultdict(lambda: EMPTY)


GameOverState = namedtuple('GameOverState', 'winner board')
