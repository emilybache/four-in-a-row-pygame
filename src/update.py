from src.constants import DROP_DELAY_MS, RED, YELLOW, ROWS, EMPTY, COLUMNS, BOARD_LEFT, BOARD_RIGHT, DISC_DIAMETER
from src.messages import LeftMouseDownAt, Tick, ColumnWasClicked, MouseMovedTo, LeftMouseUpAt
from src.states import StartScreenState, GameState, GameOverState


def update(model, msg, audio_api):
    if isinstance(model, StartScreenState):
        if isinstance(msg, LeftMouseDownAt):
            audio_api.stop_music()
            return GameState()
        if isinstance(msg, Tick):
            if not model.music_playing:
                audio_api.play_music('music')
                model.music_playing = True
            model.time = msg.time
            return model
    if isinstance(model, GameState):
        if isinstance(msg, Tick):
            model.time = msg.time
            if model.mouse_down_time:
                if model.time > model.mouse_down_time + DROP_DELAY_MS:
                    model.mouse_down_time = None
                    model = update(model, ColumnWasClicked(convert_to_column(model.mouse_pos[0])), audio_api)
        if isinstance(msg, MouseMovedTo):
            model.mouse_pos = msg.pos
        if isinstance(msg, LeftMouseDownAt):
            if convert_to_column(msg.pos[0]) is not None:
                model.mouse_down = msg.pos
                model.mouse_down_time = model.time
        if isinstance(msg, LeftMouseUpAt):
            model.mouse_down_time = None
        if isinstance(msg, ColumnWasClicked):
            audio_api.play_sound('drop')
            model.board = place_brick(model.board, model.whos_turn_is_it, msg.column)
            model.whos_turn_is_it = (model.whos_turn_is_it + 1) % 2
            for color in [RED, YELLOW]:
                won = check_winning_state(model.board, color)
                if won:
                    return GameOverState(winner=color, board=model.board)
    if isinstance(model, GameOverState):
        if isinstance(msg, LeftMouseDownAt):
            return StartScreenState()

    return model


def place_brick(board, color, column):
    log(f"Placing brick color {print_color(color)} in column {column}")
    for i in range(ROWS):
        y = ROWS - i - 1
        if board[(column, y)] == EMPTY:
            board[(column, y)] = color
            break
    return board


def print_color(color):
    return 'red' if color == RED else 'yellow'


def check_winning_state(board, color):
    for (x, y) in positions_in_print_order():
        for dir in [(0, 1), (1, 0), (1, 1), (-1, 1)]:
            cells = extract(board, (x, y), dir)
            if all(cell == color for cell in cells):
                log(f"Found 4-in-a-row at {x, y} dir {dir}")
                return True
    return False


def extract(board, pos, dir):
    cells = []
    for i in range(4):
        p = (pos[0] + dir[0] * i, pos[1] + dir[1] * i)
        cells.append(board[p])
    return cells


def positions_in_print_order():
    for y in range(ROWS):
        for x in range(COLUMNS):
            yield (x, y)


def convert_to_column(x):
    if x < BOARD_LEFT:
        return None
    if x > BOARD_RIGHT:
        return None
    return (x - BOARD_LEFT) // DISC_DIAMETER


def log(msg):
    pass
    # print(msg)