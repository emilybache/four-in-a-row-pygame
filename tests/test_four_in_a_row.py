from collections import defaultdict

import pygame
from approvaltests import verify

from src.four_in_a_row import (GameOverState, GameState, StartScreenState,
                               update, view, ColumnWasClicked,
                               LeftMouseDownAt, MouseMovedTo,
                               print_model, WIDTH, HEIGHT)
import src.four_in_a_row


class FakeDrawingApi:
    def __init__(self):
        self.surface = pygame.Surface(src.four_in_a_row.SCREENDIM)
        self.real_api = src.four_in_a_row.DrawingAPI(self.surface, '../res')

    def draw_rectangle(self, center, size, color):
        self.real_api.draw_rectangle(center, size, color)

    def draw_disc(self, center, size, color):
        self.real_api.draw_disc(center, size, color)

    def draw_text(self, center, text, size, color):
        self.real_api.draw_text(center, text, size, color)

    def draw_image(self, center, name, dimension):
        self.real_api.draw_image(center, name, dimension)


def rgb_int2tuple(rgbint):
    return (rgbint // 256 // 256 % 256, rgbint // 256 % 256, rgbint % 256)


def rgb_to_ascii(r, g, b):
    if sum([r, g, b]) < 10:
        return '.'
    if b < r > g and r > 100:
        return 'R'
    if r < g > b and g > 100:
        return 'G'
    if r < b > g and b > 100:
        return 'B'
    if r > 100 and g > 100 and b < 100:
        return 'Y'
    return 'W'


assert rgb_to_ascii(0, 0, 0) == '.'
assert rgb_to_ascii(200, 0, 0) == 'R'
assert rgb_to_ascii(0, 200, 0) == 'G'
assert rgb_to_ascii(0, 0, 200) == 'B'
assert rgb_to_ascii(200, 200, 0) == 'Y'
assert rgb_to_ascii(200, 200, 200) == 'W'


def print_result(model_and_surface):
    (model, surface) = model_and_surface
    ascii_art_width = 79
    ascii_art_height = int(ascii_art_width // (WIDTH / HEIGHT))
    smaller = pygame.transform.scale(surface, (ascii_art_width, ascii_art_height))
    ascii_art = ''
    for y in range(smaller.get_height()):
        for x in range(smaller.get_width()):
            color = smaller.get_at_mapped((x, y))
            (r, g, b) = rgb_int2tuple(color)
            ascii_color = rgb_to_ascii(r, g, b)
            ascii_art += ascii_color
        ascii_art += '\n'
    state_string = print_model(model)
    return f'''\
FINAL STATE:
{state_string}

FINAL SCREEN:
{ascii_art}

SIMULATION LOG:
{log}'''


log = ''


def fake_log(s):
    global log
    log += f"LOG: {s}\n"


src.four_in_a_row.log = fake_log


def setup_function():
    global log
    log = ''
    pygame.init()


def simulate(model, messages):
    global log
    # Mimics behaviour of main event loop in four_in_a_row
    fake_api = FakeDrawingApi()
    log += f"[SIMULATION STARTING]\n"
    log += f"===Model state===\n"
    log += f"{print_model(model)}\n\n\n"
    for msg in messages:
        log += f"[SIMULATING MSG={msg}]\n\n"
        model = update(model, msg)
        log += f"===Model state===\n"
        log += f"{print_model(model)}\n\n\n"
    log += f"[SIMULATION ENDED]"

    view(model, fake_api)
    return (model, fake_api.surface)


def test_first_placed_brick_is_red():
    result = simulate(GameState(), [ColumnWasClicked(0)])
    verify(print_result(result))


def test_placing_4_bricks_in_first_column():
    model = GameState()
    result = simulate(model, [ColumnWasClicked(0)] * 4)
    verify(print_result(result))


def test_letting_red_win():
    model = GameState()
    msgs = [ColumnWasClicked(0), ColumnWasClicked(1)] * 3 + [ColumnWasClicked(0)]
    result = simulate(model, msgs)
    verify(print_result(result))


def test_letting_red_win_horisontally():
    model = GameState()
    result = simulate(model, [ColumnWasClicked(c) for c in [0, 5, 1, 5, 2, 5, 3]])
    verify(print_result(result))


def test_letting_yellow_win_horisontally():
    model = GameState()
    result = simulate(model, [ColumnWasClicked(c) for c in [0, 1, 0, 2, 0, 3, 1, 4]])
    verify(print_result(result))


def test_slash_red_win():
    model = GameState()
    result = simulate(model, [ColumnWasClicked(c) for c in [
        0, 1,
        1, 2,
        2, 3,
        2, 3,
        3, 5,
        3]])
    verify(print_result(result))


def test_backslash_yellow_win():
    model = GameState()
    result = simulate(model, [ColumnWasClicked(c) for c in [
        0, 6,
        5, 5,
        4, 4,
        3, 4,
        5, 3,
        0, 3,
        0, 3]])
    verify(print_result(result))


def test_startscreen():
    model = StartScreenState()
    result = simulate(model, [])
    verify(print_result(result))


def test_clicking_in_game_over_state():
    print("HESAN")
    import os
    print(os.getcwd())
    model = GameOverState(winner=src.four_in_a_row.RED, board=src.four_in_a_row.empty_board())
    result = simulate(model, [LeftMouseDownAt((1, 1))])
    verify(print_result(result))


def test_startscreen_to_game_transition():
    model = StartScreenState()
    result = simulate(model, [LeftMouseDownAt((1, 1))])
    verify(print_result(result))


def test_mouse_movement_over_game_screen():
    model = GameState()
    result = simulate(model, [MouseMovedTo((x, x)) for x in range(0, 500, 10)])
    verify(print_result(result))
