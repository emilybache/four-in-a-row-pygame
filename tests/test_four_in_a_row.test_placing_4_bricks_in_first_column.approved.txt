STATE:
GameState
It is reds turn.
The mouse is at (550, 400).
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
Y O O O O O O
R O O O O O O
Y O O O O O O
R O O O O O O


LOG:
Placing brick color red in column 0
Board state is now:
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
R O O O O O O

Placing brick color yellow in column 0
Board state is now:
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
Y O O O O O O
R O O O O O O

Placing brick color red in column 0
Board state is now:
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
R O O O O O O
Y O O O O O O
R O O O O O O

Placing brick color yellow in column 0
Board state is now:
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
Y O O O O O O
R O O O O O O
Y O O O O O O
R O O O O O O

Drawing rectangle center (550, 400) size (1100, 800) color (0, 0, 0)
Drawing a disc center (550, 85) size 45 color (255, 0, 0)
Drawing rectangle center (550, 400) size (650, 560) color (0, 0, 255)
Drawing a disc center (280, 175) size 45 color (0, 0, 0)
Drawing a disc center (370, 175) size 45 color (0, 0, 0)
Drawing a disc center (460, 175) size 45 color (0, 0, 0)
Drawing a disc center (550, 175) size 45 color (0, 0, 0)
Drawing a disc center (640, 175) size 45 color (0, 0, 0)
Drawing a disc center (730, 175) size 45 color (0, 0, 0)
Drawing a disc center (820, 175) size 45 color (0, 0, 0)
Drawing a disc center (280, 265) size 45 color (0, 0, 0)
Drawing a disc center (370, 265) size 45 color (0, 0, 0)
Drawing a disc center (460, 265) size 45 color (0, 0, 0)
Drawing a disc center (550, 265) size 45 color (0, 0, 0)
Drawing a disc center (640, 265) size 45 color (0, 0, 0)
Drawing a disc center (730, 265) size 45 color (0, 0, 0)
Drawing a disc center (820, 265) size 45 color (0, 0, 0)
Drawing a disc center (280, 355) size 45 color (255, 255, 0)
Drawing a disc center (370, 355) size 45 color (0, 0, 0)
Drawing a disc center (460, 355) size 45 color (0, 0, 0)
Drawing a disc center (550, 355) size 45 color (0, 0, 0)
Drawing a disc center (640, 355) size 45 color (0, 0, 0)
Drawing a disc center (730, 355) size 45 color (0, 0, 0)
Drawing a disc center (820, 355) size 45 color (0, 0, 0)
Drawing a disc center (280, 445) size 45 color (255, 0, 0)
Drawing a disc center (370, 445) size 45 color (0, 0, 0)
Drawing a disc center (460, 445) size 45 color (0, 0, 0)
Drawing a disc center (550, 445) size 45 color (0, 0, 0)
Drawing a disc center (640, 445) size 45 color (0, 0, 0)
Drawing a disc center (730, 445) size 45 color (0, 0, 0)
Drawing a disc center (820, 445) size 45 color (0, 0, 0)
Drawing a disc center (280, 535) size 45 color (255, 255, 0)
Drawing a disc center (370, 535) size 45 color (0, 0, 0)
Drawing a disc center (460, 535) size 45 color (0, 0, 0)
Drawing a disc center (550, 535) size 45 color (0, 0, 0)
Drawing a disc center (640, 535) size 45 color (0, 0, 0)
Drawing a disc center (730, 535) size 45 color (0, 0, 0)
Drawing a disc center (820, 535) size 45 color (0, 0, 0)
Drawing a disc center (280, 625) size 45 color (255, 0, 0)
Drawing a disc center (370, 625) size 45 color (0, 0, 0)
Drawing a disc center (460, 625) size 45 color (0, 0, 0)
Drawing a disc center (550, 625) size 45 color (0, 0, 0)
Drawing a disc center (640, 625) size 45 color (0, 0, 0)
Drawing a disc center (730, 625) size 45 color (0, 0, 0)
Drawing a disc center (820, 625) size 45 color (0, 0, 0)
Drawing text 'Red to place disc' at (550, 20) color (255, 255, 255) size 30
Drawing rectangle center (235, 400) size (2, 100) color (0, 255, 0)
Drawing rectangle center (865, 400) size (2, 100) color (0, 255, 0)
